### PR TITLE
Add MATCHES features.filters settings to deprecate enableRegex

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR deprecates the `enableRegex` config option in favour of a new `MATCHES` option added to `features.filters`.

The `_MATCHES` comparator can now be enabled separately for `String` or `ID`, as follows:

```ts
const features = {
    filters: {
        String: {
            MATCHES: true,
        },
        ID: {
            MATCHES: true,
        }
    }
};

const neoSchema = new Neo4jGraphQL({ features, typeDefs, driver });
```

## Complexity

Complexity:

Medium

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-stage-3076.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 34 | 43 | 1134 |
|  aggregations.NestedAggregation | 14551 | 14551 | 87 | 103 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 64 | 102 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3304 | 3578 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2947 | 2936 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 56 | 105 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 71 | 66 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 4684 | 4866 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 123 | 159 | 600 |
|  connect.createAndConnect | 14424 | 14424 | 197 | 212 | 3003 |
|  connections.Connection | 14082 | 14082 | 58 | 186 | 2174 |
|  connections.NestedConnection | 41462 | 41462 | 117 | 137 | 4516 |
|  connections.ConnectionWithSort | 2282 | 2282 | 56 | 64 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2282 | 2282 | 158 | 253 | 1040 |
|  create.SimpleMutation | 7 | 7 | 58 | 74 | 1 |
|  cypher-directive.TopLevelCypherDirective | 116092 | 116092 | 815 | 875 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 147609 | 147609 | 277 | 348 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 357 | 310 | 1040 |
|  delete.NestedDeleteInUpdate | 22200 | 22200 | 164 | 184 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 42 | 40 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 63 | 80 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 3450 | 3450 | 22 | 26 | 1134 |
|  2925.NestedSingleRelationshipRequiredFilter | 9383 | 9383 | 43 | 75 | 1040 |
|  query.SimpleQuery | 15160 | 15160 | 59 | 78 | 2174 |
|  query.QueryWhere | 9700 | 9716 | 30 | 39 | 2162 |
|  query.SimpleQueryWithNestedWhere | 9878 | 9894 | 45 | 53 | 2162 |
|  query.Nested | 10092037 | 10092037 | 10137 | 11088 | 2008534 |
|  query.NestedWithFilter | 40401 | 40401 | 107 | 199 | 4511 |
|  query.OrFilterOnRelationships | 43701 | 41220 | 156 | 173 | 1975 |
|  query.OrFilterOnRelationshipsAndNested | 37596 | 34442 | 194 | 214 | 1975 |
|  query.QueryWithNestedIn | 13589 | 13249 | 49 | 58 | 1281 |
|  query.NestedConnectionWhere | 9834 | 9834 | 53 | 58 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9929 | 9889 | 66 | 89 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19200 | 19140 | 133 | 240 | 1614 |
|  query.NestedWithRelationshipSingleFilters | 3024400 | 3024400 | 615 | 766 | 1003942 |
|  query.Fulltext | 96 | 96 | 25 | 26 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 39 | 41 | 84 |
|  sorting.SortMultipleTypes | 2513 | 2513 | 65 | 63 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1473 | 1541 | 134 | 202 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 73 | 57 | 2174 |
|  sorting.SortDeeplyNestedFields | 42196 | 42196 | 105 | 111 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 90 | 142 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 72 | 72 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 56 | 73 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 333 | 325 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 320 | 279 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 59 | 68 | 1040 |

</details>
        
Old Schema Generation: 28.370s
Schema Generation: 28.397s
Old Subgraph Schema Generation: 42.948s
Subgraph Schema Generation: 42.537s

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-3076.surge.sh

File: packages/graphql-toolbox/src/modules/SchemaView/SchemaView.tsx [hunk 1]
```diff
@@ -99,12 +99,25 @@ export const SchemaView = ({ hasSchema, onChange }: Props) => {
 
                 Storage.storeJSON(LOCAL_STATE_TYPE_DEFS, typeDefs);
 
+                const features = isRegexChecked
```
Comment (darrellwarde) [inline packages/graphql-toolbox/src/modules/SchemaView/SchemaView.tsx:102]: Good catch this!

File: packages/graphql/src/schema/get-where-fields.ts [hunk 2]
```diff
@@ -126,17 +126,26 @@ function getWhereFields({
...
 
             if (["String", "ID"].includes(f.typeMeta.name)) {
-                if (enableRegex) {
+                const matchesSetting: boolean | undefined = features?.filters?.[f.typeMeta.name]?.MATCHES;
+                if (matchesSetting !== undefined) {
+                    if (matchesSetting === true) {
+                        res[`${f.fieldName}_MATCHES`] = { type: "String", directives: deprecatedDirectives };
+                    }
+                } else if (enableRegex) {
+                    // Deprecated. To be removed.
```
Comment (darrellwarde) [inline packages/graphql/src/schema/get-where-fields.ts:135]: I know you're probably going to do this very soon, but maybe make a proper `TODO` so extensions can find it?

File: docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc [hunk 3]
```diff
@@ -250,13 +253,31 @@ const { CypherRuntime } = require("@neo4j/graphql");
...
+| Enables MATCHES comparator.
+
 |===
 
 [[api-reference-neo4jgraphql-input-neo4jgraphqlplugins]]
 ==== `Neo4jGraphQLPlugins`
 
 |===
-|Name and Type |Description
+|Name and Type |DescriptionW
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc:280]:
```suggestion
|Name and Type |Description
```

File: docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc [hunk 4]
```diff
@@ -221,9 +221,12 @@ const { CypherRuntime } = require("@neo4j/graphql");
  +
  Type: xref::api-reference/neo4jgraphql.adoc#api-reference-neo4jgraphql-input-neo4jstringfilterssettings[`Neo4jStringFiltersSettings`]
 |Additional configuration for String filters.
+|`ID` +
+ +
+ Type: xref::api-reference/neo4jgraphql.adoc#api-reference-neo4jgraphql-input-neo4jstringfilterssettings[`Neo4jIDFiltersSettings`]
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc:226]:
```suggestion
 Type: xref::api-reference/neo4jgraphql.adoc#api-reference-neo4jgraphql-input-neo4jidfilterssettings[`Neo4jIDFiltersSettings`]
```

File: packages/graphql/src/schema/get-where-fields.ts [hunk 5]
```diff
@@ -126,17 +126,26 @@ function getWhereFields({
...
                 }
 
                 const stringWhereOperators = ["_CONTAINS", "_STARTS_WITH", "_ENDS_WITH"];
 
                 const stringWhereOperatorsNegate = ["_NOT_CONTAINS", "_NOT_STARTS_WITH", "_NOT_ENDS_WITH"];
 
-                Object.entries(features?.filters?.String || {}).forEach(([key, value]) => {
-                    if (value) {
-                        stringWhereOperators.push(`_${key}`);
+                Object.entries(features?.filters?.String || {}).forEach(([filter, enabled]) => {
```
Comment (darrellwarde) [inline packages/graphql/src/schema/get-where-fields.ts:143]: Could we not use this loop to add the `MATCHES` filter, especially on the 4.0.0 branch when `enableRegex` is out of the way?

**Review by darrellwarde — APPROVED** [review]
It all looks good, ready to merge following a couple of tweaks!

File: packages/graphql/src/schema/get-where-fields.ts (hunk 5 above)
Comment (mjfwebb) [reply packages/graphql/src/schema/get-where-fields.ts:143]: I definitely want to do this but I would like to do it in the `4.0.0` branch as you say.

File: packages/graphql/src/schema/get-where-fields.ts (hunk 2 above)
Comment (mjfwebb) [reply packages/graphql/src/schema/get-where-fields.ts]: Updated it in https://github.com/neo4j/graphql/pull/3076/commits/a5802bb84acb1c02ae25f9abcbe6830942d0b460

**Review by darrellwarde — APPROVED** [review]

**Pull request merged**